### PR TITLE
`Communication`: Improve unread messages count view on sidebar

### PR DESCRIPTION
**Pull request by asliayk**
#### General

- [X] This is a small issue that I tested locally.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [X] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [X] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [X] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [X] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Previously, the sidebar did not display the unread message count for the relevant conversations. This update refactors the view to show the number of unread messages next to each conversation in the sidebar.

### Description

The unread count is now displayed next to the conversation title, with a special format (99+) for counts exceeding 99. Additionally, the conversation title is now displayed in bold when there are unread messages, enhancing the visibility of conversations that require attention. 

### Steps for Testing

Prerequisites:
- 2 Instructors/Students
- 1 Course with communication enabled

1. Log in to Artemis
2. Navigate to Communication section of a course
3. Send arbitrary number of messages to different channels
4. Log in to Artemis with another account, navigate to the communication section of the same course
5. Notice the new view of unread messages count next to the related conversation titles

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| sidebar-card-item.component.ts | 100% | ✅ |

### Screenshots

image
image

## Release Notes

- **New Features**
	- Enhanced sidebar rendering logic for 'exam' and 'exercise' types, improving visual cues for unread notifications.
	- Introduced a new CSS class for unread message indicators.
	- Added support for displaying unread message counts in sidebar cards across different components.

- **Bug Fixes**
	- Improved handling of unread message counts to ensure accurate display.

- **Tests**
	- Added unit tests for verifying the functionality of unread message formatting in the sidebar component.
  
- **Style**
	- Updated dropdown toggle button styling for improved visual presentation.

### CODE REVIEW

File: src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html [hunk 1]
```diff
@@ -58,6 +64,9 @@
                         <fa-icon [fixedWidth]="true" [icon]="sidebarItem.rightIcon" />
                     }
                 </span>
+                @if (unreadCount && unreadCount > 0) {
+                    <span class="unread-count">{{ formattedUnreadCount }}</span>
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html:68]: I would decrease the spacing between the heart and the counter: `me-n2`
In addition I would also add `px-2` to the button of the `faEllipsisVertical `icon in the `conversation-options.component.html` file

**Review by rabeatwork — CHANGES_REQUESTED** [review]
I have some remarks.

File: src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.ts [hunk 2]
```diff
@@ -1,13 +1,27 @@
...
     selector: 'jhi-sidebar-card-item',
     templateUrl: './sidebar-card-item.component.html',
     styleUrls: ['./sidebar-card-item.component.scss', '../sidebar.component.scss'],
 })
-export class SidebarCardItemComponent {
+export class SidebarCardItemComponent implements OnInit {
     @Input() sidebarItem: SidebarCardElement;
     @Input() sidebarType?: SidebarTypes;
     @Input() groupKey?: string;
+    @Input() unreadCount: number = 0;
```
Comment (PaRangger) [inline src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.ts:13]: I think we should use input signals over @Input() according to the docs.

```suggestion
    unreadCount = input<number>(0);
```

File: src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.ts [hunk 3]
```diff
@@ -1,13 +1,27 @@
...
+
+    formattedUnreadCount: string = '';
+
+    ngOnInit(): void {
```
Comment (PaRangger) [inline src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.ts:17]: I'm unsure if its due something else, but formatting the number only on initialization of the component means the numbers do not respond to the websocket notifications. Maybe it is possible to react to changes of unreadCount, formatting again and displaying the new number.

Comment (asliayk) [reply src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.ts:17]: you're right, thank you. I added the related logic to `ngOnChanges` and I think it solves this issue now.

**Review by krusche — APPROVED** [review]
Changes look good to me 👍 and are a great improvements for users 🥇

**Review by PaRangger — APPROVED** [review]
Thanks for the changes. 😊 Tested on TS4, code LGTM. 👍🏼

**Review by az108 — APPROVED** [review]
Tested on TS4, works without any issues. Works also for different channel groups and categories. 👍 

image

**Review by SimonEntholzer — APPROVED** [review]
Works on Ts4, when adding new messages, the number of new messages shows up, nice improvement :+1:
Two remarks:
- should the number disappear after I clicked on the channel? I think at them moment there is some kind of timer?
- also, when I add messages myself, should I see the new message count go up?

What would also be nice in a follow up, is to directly see new incoming messages, without needing to click somewhere to update the channels :D

**Comment by az108** [comment]
I also noticed the second point made by @SimonEntholzer . In my opinion it could be considered removing this as it is something uncommon

**Review by cremertim — APPROVED** [review]

**Pull request merged**